client/internal/providerqueryer/requests.go

```diff
@@ -0,0 +1,163 @@
+package providerqueryer
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+)
+
+type rwLocker interface {
+	sync.Locker
+	RLock()
+	RUnlock()
+}
+
+// for overriding in tests
+var newMutex = func() sync.Locker { return &sync.Mutex{} }
+var newRWMutex = func() rwLocker { return &sync.RWMutex{} }
```

Comment:
You are adding a virtual call and a memory allocation just to override mutex in tests ? Are you sure it's worth it ?

Reply:
catching deadlocks is pretty important, I'd rather discover later that this is a bottleneck